File: atomic/src/main/java/com/infideap/atomic/AtomRequest.java
package com.infideap.atomic;

import android.os.Handler;
import android.os.Looper;

import java.io.IOException;
import java.lang.reflect.Type;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * Created by Shiburagi on 20/10/2016.
 */
public class AtomRequest<T> {
    private static final String TAG = AtomRequest.class.getSimpleName();
    private final A a;
    private final Class<?> aClass;

    protected T t = null;

    AtomRequest(A a, Class<T> aClass) {
        this.a = a;
        this.aClass = aClass;

        a.createClient();
    }

    AtomRequest(A a, Type t) {
        this.a = a;
        this.aClass = t.getClass();

        a.createClient();
    }


    public void setCallback(final FutureCallback<T> callback) {


        prepare().enqueue(new Callback() {
            @Override
            public void onFailure(Call call, final IOException e) {
                Handler handler = new Handler(Looper.getMainLooper());
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        onCompleted(e, t, null);
                    }
                });
            }


            @Override
            public void onResponse(Call call, final Response response) {
                Exception exception;
                try {
                    parse(response);
                    exception = null;
                } catch (Exception e) {
                    exception = e;
                }
                final Exception e = exception;
                Handler handler = new Handler(Looper.getMainLooper());
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        onCompleted(e, t, response);
                    }
                });
            }

            private void onCompleted(Exception e, T t, Response response) {
                if (callback instanceof ResponseFutureCallback) {
                    ((ResponseFutureCallback) callback).onCompleted(e, t, response);
                } else
                    callback.onCompleted(e, t);
            }


        });

    }

    public T get() throws IOException {
        return parse(prepare().execute());
    }

    protected T parse(Response response) throws IOException {
        if (a.isString) {
            t = (T) response.body().string();
        } else {
            t = (T) a.gson.fromJson(response.body().charStream(), aClass);
        }

        try {
            response.body().close();
        } catch (Exception ignored) {
        }
        return t;


    }

    protected Call prepare() {
        RequestBody requestBody = RequestBody.create(A.JSON, a.getRequestBody() == null ? "" : a.getRequestBody().toString());
        if (a.getParts() != null) {
            if (a.getParts().size() > 0) {
                MultipartBody.Builder builder = new MultipartBody.Builder()
                        .setType(MultipartBody.FORM);
                for (Part part : a.getParts()) {

                    MediaType mediaType = MediaType.parse("application/octet-stream");
                    if (part instanceof FilePart) {
                        FilePart filePart = (FilePart) part;
                        if (filePart.getFile() == null)
                            continue;
                        RequestBody body = RequestBody.create(mediaType, filePart.getFile());
                        builder = builder.addFormDataPart(part.getKey(), filePart.getName(), body);
                    }else{
                        builder = builder.addFormDataPart(part.getKey(), part.getName());
                    }
                }
                if (a.getRequestBody() != null)
                    builder.addPart(requestBody);
                requestBody = builder.build();
            }
        }
        Request.Builder builder = new Request.Builder()
                .url(a.getUrl());

        if (a.headers != null)
            builder.headers(a.headers);

        Request request = a.method == null ? (
                a.getRequestBody() == null ?
                        builder.get().build() :
                        builder.post(requestBody).build()
        ) : builder.method(a.method.toUpperCase(), requestBody).build();
        a.request = request;


        return a.client.newCall(request);
    }


    Response execute() throws IOException {
        return prepare().execute();
    }
}

File: atomic/src/main/java/com/infideap/atomic/AtomRequest.kt
package com.infideap.atomic

import android.os.Handler
import android.os.Looper
import com.infideap.atomic.AtomRequest
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import java.io.IOException
import java.lang.reflect.Type

/**
 * Created by Shiburagi on 20/10/2016.
 */
open class AtomRequest<T> internal constructor(private val a: A, aClass: Class<T>?) {
    private val aClass: Class<*>? = aClass
    protected var t: T? = null

    init {
        a.createClient()
    }

//    internal constructor(a: A, t: Type) {
//        this.a = a
//        aClass = t.javaClass
//        a.createClient()
//    }

    fun notNull(t:T?):T{
        return t?:aClass?.newInstance() as T;
    }

    open fun setCallback(callback: FutureCallback<T>) {
        prepare().enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                val handler = Handler(Looper.getMainLooper())
                handler.post { onCompleted(e, t, null) }
            }

            override fun onResponse(call: Call, response: Response) {
                val exception: Exception?
                exception = try {
                    parse(response)
                    null
                } catch (e: Exception) {
                    e
                }
                val handler = Handler(Looper.getMainLooper())
                handler.post { onCompleted(exception, t, response) }
            }

            private fun onCompleted(e: Exception?, t: T?, response: Response?) {
                if (callback is ResponseFutureCallback<T>) {
                    callback.onCompleted(e, notNull(t), response)
                } else callback.onCompleted(e, notNull(t))
            }
        })
    }

    @Throws(IOException::class)
    open fun get(): T? {
        return parse(prepare().execute())
    }

    @Throws(IOException::class)
    protected open fun parse(response: Response): T? {
        t = if (a.isString) {
            response.body!!.string() as T
        } else {
            a.gson.fromJson(response.body!!.charStream(), aClass) as T
        }
        try {
            response.body!!.close()
        } catch (ignored: Exception) {
        }
        return t
    }

    protected fun prepare(): Call {
        var requestBody = RequestBody.create(A.Companion.JSON, if (a.requestBody == null) "" else a.requestBody.toString())
        if (a.parts != null) {
            if (a.parts!!.size > 0) {
                var builder = MultipartBody.Builder()
                        .setType(MultipartBody.FORM)
                for (part in a.parts!!) {
                    val mediaType: MediaType = "application/octet-stream".toMediaTypeOrNull()!!
                    builder = if (part is FilePart) {
                        val filePart = part
                        if (filePart.file == null) continue
                        val body = RequestBody.create(mediaType, filePart.file!!)
                        builder.addFormDataPart(part.key, filePart.name, body)
                    } else {
                        builder.addFormDataPart(part.key, part.name?:"")
                    }
                }
                if (a.requestBody != null) builder.addPart(requestBody)
                requestBody = builder.build()
            }
        }
        val builder = Request.Builder()
                .url(a.url?:"")
        if (a.headers != null) builder.headers(a.headers!!)
        val request = if (a.method == null) if (a.requestBody == null) builder.get().build() else builder.post(requestBody).build() else builder.method(a.method!!.toUpperCase(), requestBody).build()
        a.request = request
        return a.client!!.newCall(request)
    }

    @Throws(IOException::class)
    fun execute(): Response {
        return prepare().execute()
    }

    companion object {
        private val TAG = AtomRequest::class.java.simpleName
    }
}